New OP-TEE SHM design.

lorc opened the pull request:
Those patches are kernel driver part of dynamic shared memory in OP-TEE (refer to https://github.com/OP-TEE/optee_os/pull/1232). Two of those patches are authored by @jenswi-linaro, I just tidied it a bit and rebased on current master.

This patches provide new ioctl to userpace along with support in tee_shm and op-tee kernel part. This ioctl allows userspace process to share own memory with OP-TEE. 

Another part - is OP-TEE specific SHM pool that uses kernel memory. With this approach we don't rely on OP-TEE predefined pool.

jenswi-linaro committed:
[NEW SHM]tee: add register user memory

Added new ioctl to allow users register own buffers as a shared memory.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>
Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: shm: add accessors for buffer size and page offset

These two function will be needed for shared memory registration in OP-TEE

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

etienne-lms commented on drivers/tee/tee_core.c:
```diff
@@ -146,6 +146,45 @@ static int tee_ioctl_shm_alloc(struct tee_context *ctx,
+	data.id = -1;
+
+	shm = tee_shm_register(ctx, data.addr, data.length,
+			       TEE_SHM_DMA_BUF|TEE_SHM_USER_MAPPED);
```
minor: spaces between `|`

etienne-lms commented on drivers/tee/tee_core.c:
```diff
@@ -146,6 +146,45 @@ static int tee_ioctl_shm_alloc(struct tee_context *ctx,
+	if (data.flags)
+		return -EINVAL;
+
+	data.id = -1;
```
minor: why this? (same question applies to `tee_ioctl_shm_alloc() ` above.

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -190,9 +194,13 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+		mutex_lock(&teedev->mutex);
+		list_add_tail(&shm->link, &ctx->list_shm);
+		mutex_unlock(&teedev->mutex);
+
```
minor: empty line

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -207,8 +215,136 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+	void *ret;
+	int rc;
+	int num_pages;
+	size_t n;
```
minor: `n` could move to loop.

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -207,8 +215,136 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+				 size_t length, u32 flags)
+{
+	struct tee_device *teedev = ctx->teedev;
+	u32 req_flags = TEE_SHM_DMA_BUF|TEE_SHM_USER_MAPPED;
```
minor: spaces between `|`

etienne-lms reviewed (commented):
late feedback.
few minors and an error case to crosscheck.
very nice.

etienne-lms commented on drivers/tee/optee/call.c:
```diff
@@ -433,3 +437,125 @@ void optee_disable_shm_cache(struct optee *optee)
+		mp->attr |= OPTEE_MSG_ATTR_FRAGMENT;
+		mp++;
+		/* Check if we filled whole page with addresses */
+		if (((uintptr_t)(mp + 1) & ~PAGE_MASK) == 0) {
```
i don't catch this.
should not be `((uintptr_t)mp & PAGE_MASK) != (uintptr_t)(mp + 1) & PAGE_MASK)` ?

lorc commented:
Addressed @etienne-lms comments, rebased on actual branch.

jenswikl commented on include/linux/tee_drv.h:
```diff
@@ -286,6 +317,13 @@ int tee_shm_get_id(struct tee_shm *shm);
+/**
+ * tee_shm_is_registered() - Check if shared memory object in registered in TEE
+ * @shm:	Shared memory handle
+ * @returns true if object is registed in TEE
```
`s/registed/registered/`

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -433,3 +437,126 @@ void optee_disable_shm_cache(struct optee *optee)
 	optee_cq_wait_final(&optee->call_queue, &w);
 }
+
+size_t optee_correct_params_cnt(size_t num_params)
```
I find this name confusing. What about using `optee_round_up_params_count()` instead?

jenswi-linaro reviewed (commented):
Why is an own shared memory pool needed? It does more or less the same thing as the pool in the subsystem.

lorc commented:
Rebased, addressed comments. Also added new patch (for privileged calls) and simplified RPC memory allocation.

@jenswi-linaro, own shared memory pool needed because I want to use kernel memory for command buffers.

jenswi-linaro commented:
@lorc, `kmalloc()` should be able to supply the memory needed for command data and `vmalloc()` for the larger allocation, or am I missing something?

lorc commented:
@jenswi-linaro, you are right. Actually, "new shared memory pool" is not a right name for it. But it is the term used in tee code. It is more like an allocator. New "pool" uses `kmalloc()` to allocate command buffers. And it uses OP-TEE shared memory region to allocate dma_buf regions, so `TEE_IOC_SHM_ALLOC` will allocate memory from that plain old 2M buffer. It was left for compatibility because new clients will not use `TEE_IOC_SHM_ALLOC`.

lorc committed:
[NEW SHM]tee: shm: add page accessor functions

In order to register a shared buffer in TEE, we need accessor
function that return list of pages for that buffer.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc commented:
Hi. I just rebased this on actual branch and also added delta patch with new ABI and new RPC cleanup approach.

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +453,124 @@ void optee_disable_shm_cache(struct optee *optee)
+}
+
+/* Number of user pages + number of pages to hold list of user pages  */
+#define GET_ARRAY_SIZE(num_entries) \
```
Declare this as a static function instead.

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +453,124 @@ void optee_disable_shm_cache(struct optee *optee)
+
+	msg_arg->cmd = OPTEE_MSG_CMD_REGISTER_SHM;
+	msg_arg->params->attr = OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT |
+		OPTEE_MSG_ATTR_NONCONTIG;
```
Align with beginning of `OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT` above.

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +453,124 @@ void optee_disable_shm_cache(struct optee *optee)
+	msg_arg->params->u.tmem.shm_ref = (unsigned long)shm;
+	msg_arg->params->u.tmem.size = tee_shm_get_size(shm);
+	msg_arg->params->u.tmem.buf_ptr = virt_to_phys(pages_array) |
+		tee_shm_get_page_offset(shm);
```
Align with `virt_to_phys()` above.

jenswikl commented on drivers/tee/optee/core.c:
```diff
@@ -98,13 +100,79 @@ int optee_from_msg_param(struct tee_param *params, size_t num_params,
+		case OPTEE_MSG_ATTR_TYPE_RMEM_OUTPUT:
+		case OPTEE_MSG_ATTR_TYPE_RMEM_INOUT:
+			p->attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT +
+				attr - OPTEE_MSG_ATTR_TYPE_RMEM_INPUT;
```
Alignment

jenswikl commented on drivers/tee/optee/core.c:
```diff
@@ -98,13 +100,79 @@ int optee_from_msg_param(struct tee_param *params, size_t num_params,
+			/* Check that the memref is covered by the shm object */
+			if (p->u.memref.size) {
+				size_t o = p->u.memref.shm_offs +
+					p->u.memref.size;
```
Alignment

jenswikl commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -145,8 +149,9 @@ struct optee_msg_param_value {
- * OPTEE_MSG_ATTR_TYPE_RMEM_* indicates rmem.
+ * OPTEE_MSG_ATTR_TYPE_TMEM_* indicates @tmem and
+ * OPTEE_MSG_ATTR_TYPE_RMEM_* indicates @rmem,
+ * OPTEE_MSG_ATTR_TYPE_NESTED indicates @nested parameters.
```
Nested isn't used

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -201,14 +203,17 @@ static struct tee_shm *cmd_alloc_suppl(struct tee_context *ctx, size_t sz)
 	size_t sz;
 	size_t n;
 
+	BUG_ON(!call_ctx || call_ctx->pages_array || call_ctx->num_entries);
```
This shouldn't be needed

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -247,17 +252,51 @@ static void handle_rpc_func_cmd_shm_alloc(struct tee_context *ctx,
+		call_ctx->num_entries = page_num;
+
+		arg->params[0].attr =  OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT |
+			OPTEE_MSG_ATTR_NONCONTIG;
```
Alignment

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -247,17 +252,51 @@ static void handle_rpc_func_cmd_shm_alloc(struct tee_context *ctx,
+		arg->params[0].attr =  OPTEE_MSG_ATTR_TYPE_TMEM_OUTPUT |
+			OPTEE_MSG_ATTR_NONCONTIG;
+		arg->params[0].u.tmem.buf_ptr = virt_to_phys(pages_array)
+			+ tee_shm_get_page_offset(shm);
```
`+` should be at the end of the line above
alignment

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -375,16 +430,19 @@ static void handle_rpc_func_cmd(struct tee_context *ctx, struct optee *optee,
 		break;
 	case OPTEE_MSG_RPC_CMD_SHM_ALLOC:
-		handle_rpc_func_cmd_shm_alloc(ctx, arg);
+		free_page_array(call_ctx);
```
Contrary to what I've said before, it should be enough to call this function here (and of course from `optee_rpc_finalize_call()`) as not much is gained by calling it for all the other RPCs.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -409,6 +550,55 @@ void *tee_shm_get_va(struct tee_shm *shm, size_t offs)
+ *
+ * This function is not thread safe and have no reference counter
+ */
+void *tee_shm_vmap(struct tee_shm *shm)
```
Why do we need this function?

lorc commented:
Addressed comments (and also `checkpatch` issues). Removed function `tee_shm_vmap()` with friends (it was used in prev. ABI version).

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -480,21 +480,22 @@ void optee_fill_pages_list(u64 *dst, struct page **pages, size_t num_pages)
 /* Number of user pages + number of pages to hold list of user pages  */
-#define GET_ARRAY_SIZE(num_entries) \
-	sizeof(u64) * (num_entries + (sizeof(u64) * num_entries) / PAGE_SIZE)
+size_t get_pages_array_size(size_t num_entries)
```
`static`

lorc commented:
added `v3 delta` patch, where  `get_pages_array_size()` was made `static` function

jenswikl commented on drivers/tee/optee/optee_private.h:
```diff
@@ -149,11 +161,23 @@ int optee_cancel_req(struct tee_context *ctx, u32 cancel_id, u32 session);
 int optee_to_msg_param(struct optee_msg_param *msg_params, size_t num_params,
 		       const struct tee_param *params);
 
+void *optee_allocate_pages_array(size_t num_entries);
```
This should return a `u64 *` instead as that's what it's used for and `num_entries` also says how many `u64` the returned array should be able to hold.

jenswikl commented on drivers/tee/optee/core.c:
```diff
@@ -393,7 +459,11 @@ optee_config_shm_memremap(optee_invoke_fn *invoke_fn, void **memremaped_shm)
-	pool = tee_shm_pool_alloc_res_mem(&priv_info, &dmabuf_info);
+	/* If OP-TEE can work with unregistered SHM, we will use own pool */
+	if (sec_caps & OPTEE_SMC_SEC_CAP_DYNAMIC_SHM)
+		pool = optee_shm_get_pool(&dmabuf_info);
```
We don't need this new "own pool" just because OP-TEE can work with unregistered shared memory.
I'd rather remove it from this pull request.

lorc replied:
So you suggest to move it to an another PR? Because we need some mechanism to allocate command buffers.

jenswikl replied:
Aha, sorry I didn't realize that the already present pool doesn't allocate command buffers correctly. Please keep the "own pool".

In the meantime I'll try to figure out how this can work together with the already present pool instead of replacing it with partially duplicated code.

jenswikl replied:
I've created a pull request (https://github.com/lorc/linux/pull/1) into this branch making pool creation more flexible.

jenswikl commented on drivers/tee/optee/optee_private.h:
```diff
@@ -137,7 +149,7 @@ int optee_supp_recv(struct tee_context *ctx, u32 *func, u32 *num_params,
 		    struct tee_param *param);
 
-u32 optee_do_call_with_arg(struct tee_context *ctx, phys_addr_t parg);
+u32 optee_do_call_with_arg(struct tee_device *teedev, phys_addr_t parg);
```
I regret these `teedev` instead of ``ctx` and `optee` instead of `ctx` changes. They mostly add a bunch of unrelated changes that doesn't help much. Wouldn't you agree?

lorc replied:
Yeah, I agree. I'll look on how to get it back. AFAIR, there was some reason for that change.

jenswi-linaro reviewed (commented):
I wouldn't mind if this pull request had all the commits squashed and cleaned up either in this pull request or in a new pull request replacing this.

I'd like this to be possible to upstream more or less as is when we finally merge it.

jenswi-linaro committed:
[NEW SHM]tee: flexible shared memory  pool creation

Makes creation of shm pools more flexible by adding new more primitive
functions to allocate a shm pool. This makes it easier to add driver
specific shm pool management.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

lorc committed:
[NEW SHM]tee: optee: Update protocol definitions

There were changes in REE<->OP-TEE ABI recently.
Now ABI allows us to pass non-contiguous memory buffers as list of
pages to OP-TEE. This can be achieved by using new parameter attribute
OPTEE_MSG_ATTR_NONCONTIG.

OP-TEE also is able to use all non-secure RAM for shared buffers. This
new capability is enabled with OPTEE_SMC_SEC_CAP_DYNAMIC_SHM flag.

This patch adds necessary definitions to the protocol definition files at
Linux side.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: add page list manipulation functions

These functions will be used to pass information about shared
buffers to OP-TEE.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: add shared buffer registration functions

This change adds ops for shm_(un)register functions in tee interface.
Client application can use these functions to (un)register an own shared
buffer in OP-TEE address space. This allows zero copy data sharing between
Normal and Secure Worlds.

Please note that while those functions were added to optee code,
it does not report to userspace that those functions are available.
OP-TEE code does not set TEE_GEN_CAP_REG_MEM flag. This flag will be
enabled only after all other features of dynamic shared memory will be
implemented in subsequent patches.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: add registered shared parameters handling

Now, when client applications can register own shared buffers in OP-TEE,
we need to extend ABI for parameter passing to/from OP-TEE.

So, if OP-TEE core detects that parameter belongs to registered shared
memory, it will use corresponding parameter attribute.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: add registered buffers handling into RPC calls

With latest changes to OP-TEE we can use any buffers as a shared memory.
Thus, it is possible for supplicant to provide part of own memory
when OP-TEE asks to allocate a shared buffer.

This patch adds support for such feature into RPC handling code.
Now when OP-TEE asks supplicant to allocate shared buffer, supplicant
can use TEE_IOC_SHM_REGISTER to provide such buffer. RPC handler is
aware of this, so it will pass list of allocated pages to OP-TEE.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: store OP-TEE capabilities in private data

Those capabilities will be used in subsequent patches.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: add optee-specific shared pool implementation

This is simple pool that uses kernel page allocator. This pool can be
used in case OP-TEE supports dynamic shared memory.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc committed:
[NEW SHM]tee: optee: enable dynamic SHM support

Previous patches added various features that are needed for dynamic SHM.
Dynamic SHM allows Normal World to share any buffers with OP-TEE.
While original design suggested to use pre-allocated region (usually of
1M to 2M of size), this new approach allows to use all non-secure RAM for
command buffers, RPC allocations and TA parameters.

This patch checks capability OPTEE_SMC_SEC_CAP_DYNAMIC_SHM. If it was set
by OP-TEE, then kernel part of OP-TEE will use kernel page allocator
to allocate command buffers. Also it will set TEE_GEN_CAP_REG_MEM
capability to tell userspace that it supports shared memory registration.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc commented:
@jenswi-linaro, thank you for your pull request. I used your patches for `tee` as is, but I had to squash your patches for `shm_pool` into mine ones, because I changed patches layout. As you can see, now there are a lot more patches. I tried to separate different parts of this big feature into separate patches.

But now I'm facing the problem. As you suggested, I reverted that part for your patch that changed `ctx` to `teedev`. Problem is that I can't unregister shared buffer because it's context is being erased on close:

```
static void teedev_close_context(struct tee_context *ctx)
{
	struct tee_shm *shm;

	ctx->teedev->desc->ops->release(ctx);
	mutex_lock(&ctx->teedev->mutex);
	list_for_each_entry(shm, &ctx->list_shm, link)
		shm->ctx = NULL; <====== Right there
	mutex_unlock(&ctx->teedev->mutex);
	tee_device_put(ctx->teedev);
	kfree(ctx);
}

```
So, it all is fails with this backtrace:

```
[    4.712751] [<ffff000008774f80>] tee_shm_alloc+0x10/0x20
[    4.712860] [<ffff000008775ba4>] get_msg_arg+0x34/0xd0
[    4.712962] [<ffff000008776610>] optee_shm_unregister+0x28/0xb0
[    4.713070] [<ffff0000087744d0>] tee_shm_do_release+0xd8/0xf8
[    4.713174] [<ffff000008774558>] tee_shm_release+0x68/0x80
[    4.713275] [<ffff000008774580>] tee_shm_op_release+0x10/0x18
[    4.713382] [<ffff00000853ab58>] dma_buf_release+0x60/0x1a0
[    4.713508] [<ffff0000081daad4>] __fput+0x8c/0x1d0
```

Because `tee_shm_alloc()` receives null context.

I think, this is a reason, why you used `teedev` instead of `ctx` in the first place.

Looks like we need to add reference counter to `tee_ctx` to make it live until last `shm` will be freed.  Is this right approach?

jenswi-linaro commented:
Aha, yes I think using a reference counter is more clear.

lorc committed:
[NEW SHM]tee: use reference counting for tee_context

We need to ensure that tee_context is present until last
shared buffer will be freed.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc commented:
Hmm, looks like it is harder, than I expected. Problem is that `optee_release()` also uses `shm` objects. This increases/decreases reference count in `tee_context` which leads to recursion during `tee_context` release.
I employed a dirty hack to overcome this issue. You can see it in this patch: https://github.com/lorc/linux/commit/8bea4e4316366e65be2aeb9d9797488c8b021435
So, currently I'm looking for a proper solution. 
Actually, I can call `optee_release()` from `teedev_close_context()` as it was done earlier. Then there will be no recursion. But I'm not sure if this is semantically correct: we are telling TEE driver, that it can release context, but then we are using this context to unregister shared memory buffer.

lorc commented:
Okay, I added flag to avoid recursion: https://github.com/lorc/linux/commit/108bb599655dfc244f85ffcb8816590d1603e047
At least this does not look as ugly, as the first approach. Maybe I should return `kref` back, instead of raw atomics usage.

etienne-lms commented on include/linux/tee_drv.h:
```diff
@@ -42,11 +43,15 @@ struct tee_shm_pool;
 	struct list_head list_shm;
 	void *data;
+	struct kref refcount;
+	bool releasing;
```
I am not sure this `releasing` flag protects from anything. 
Maybe it could be used to warn that something ugly is about to be proceed.

jenswikl replied:
Please explain why this flag is needed. As I understand each `put()` is preceded by a `get()`.

lorc replied:
@jenswi-linaro, I described that in this and next comment: https://github.com/linaro-swg/linux/pull/29#issuecomment-317473581

jenswikl replied:
Sorry I forgot. Please add a comment as a reminder.

lorc replied:
It is described in comments to the structure (`@releasing`). I also extended this description with more details. Should I add extra comment before `bool releasing` ?

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -154,13 +166,13 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
 	shm->ctx = ctx;
 	if (flags & TEE_SHM_DMA_BUF)
-		poolm = &teedev->pool->dma_buf_mgr;
+		poolm = teedev->pool->dma_buf_mgr;
```
This should check if `teedev->pool->dma_buf_mgr != NULL`, as the pool registration api allows registering NULL pools.

Same below for the private pool.

~~Discard this comment as your latest API enforces both pool managers are defined.~~
Edited again :( Indeed API allows to register one or none of the pool managers.

lorc replied:
`tee_shm_pool_alloc()` checks both `priv_mgr` and `dmabuf_mgr` before registration.

etienne-lms replied:
`tee_shm_pool_alloc()` checks `(!mgr || mgr...)` hence null  `priv_mgr`/`dmabuf_mgr` are allowed.

lorc replied:
Ah, yes... You are right. 
I can change this check to `(mgr && mgr...)`. @jenswi-linaro, what do you think?

jenswikl replied:
I don't see what's useful with allowing a `NULL` `priv_mgr/dmabuf_mgr` and it also seems to complicate things.
How about updating `check_mgr_ops()` to require `mgr` to be a valid pointer?

lorc replied:
Yep. This is exactly what I thought about. Then, I'll do this.

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -56,13 +56,21 @@ static void tee_shm_release(struct tee_shm *shm)
 	} else if (shm->flags & TEE_SHM_REGISTER) {
+		int n;
 		int rc = teedev->desc->ops->shm_unregister(shm->ctx, shm);
-
```
minor: keep empty line

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -190,9 +202,13 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+		mutex_lock(&teedev->mutex);
+		list_add_tail(&shm->link, &ctx->list_shm);
+		mutex_unlock(&teedev->mutex);
+		teedev_ctx_get(ctx);
```
minor: get the ctx before using it ? (before `list_add_tail()` or even before call of `rc = poolm->ops->alloc(poolm, shm, size);` above)

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -207,8 +223,141 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+				 size_t length, u32 flags)
+{
+	struct tee_device *teedev = ctx->teedev;
+	u32 req_flags = TEE_SHM_DMA_BUF | TEE_SHM_USER_MAPPED;
```
minor: `const u32 req_flags = ...`

etienne-lms commented on include/uapi/linux/tee.h:
```diff
@@ -146,6 +147,36 @@ struct tee_ioctl_shm_register_fd_data {
 				     struct tee_ioctl_shm_register_fd_data)
 
 /**
+ * struct tee_shm_register_data - Shared memory register argument
```
s/tee_shm_register_data/tee_ioctl_shm_register_data/

lorc commented:
Ping?

jenswi-linaro commented:
I'm just back from vacation, I'll resume reviewing this.

etienne-lms reviewed (commented):
Some old pending comment of mine...

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -207,8 +223,141 @@ struct tee_shm *tee_shm_alloc(struct tee_context *ctx, size_t size, u32 flags)
+	if (shm) {
+		size_t n;
+
+		if (shm->id > 0) {
```
`if (shm->id >= 0)`

etienne-lms commented on drivers/tee/tee_shm.c:
```diff
@@ -427,6 +579,42 @@ int tee_shm_get_pa(struct tee_shm *shm, size_t offs, phys_addr_t *pa)
+{
+	return shm->size;
+}
+EXPORT_SYMBOL_GPL(tee_shm_get_size);
```
minor: inlined into tee_drv.h ? 
Same  comment for `tee_shm_get_page_offset()`, `tee_shm_get_pages()` and `tee_shm_is_registered()`.

etienne-lms commented on drivers/tee/tee_shm_pool.c:
```diff
@@ -104,42 +73,111 @@ struct tee_shm_pool *
+
+	/*
+	 * Start and end must be page aligned
+	 */
```
minor stuf:
- single line comment here
- below: can remove inner parenthesis.
- above: **`const`**`size_t page_mask = PAGE_SIZE - 1;`

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -46,17 +46,31 @@ static void tee_shm_release(struct tee_shm *shm)
+	} else if (shm->flags & TEE_SHM_REGISTER) {
+		int n;
+		int rc = teedev->desc->ops->shm_unregister(shm->ctx, shm);
+		if (rc)
```
Add an empty line after declaration of local variables.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -46,17 +46,31 @@ static void tee_shm_release(struct tee_shm *shm)
 
 		poolm->ops->free(poolm, shm);
+	} else if (shm->flags & TEE_SHM_REGISTER) {
+		int n;
```
Use an unsigned type, preferably `size_t` instead as `shm->num_pages` is a `size_t`.

jenswikl commented on drivers/tee/tee_shm.c:
```diff
@@ -427,6 +579,42 @@ int tee_shm_get_pa(struct tee_shm *shm, size_t offs, phys_addr_t *pa)
+ * @shm:	Shared memory handle
+ * @returns size of shared memory
+ */
+ssize_t tee_shm_get_size(struct tee_shm *shm)
```
Could be an inline function instead as `struct tee_shm` now is declared in `<linux/tee_drv.h>`.
 Also, I don't see the benefit of using `ssize_t` over the expected `size_t` type here.
Please check/update the other small access functions added in the PR too.

jenswi-linaro reviewed (commented):
For commit message "optee: add shared buffer registration functions"
```
        s/does not reports/does not report/
        s/does not sets/does not set/
```

Also I think all the commits prefixed by only "optee:" should be prefixed with "tee: ".

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +450,125 @@ void optee_disable_shm_cache(struct optee *optee)
+ * optee_fill_pages_list() - write list of user pages to given shared
+ * buffer.
+ *
+ * @dst - page-aligned buffer where list of pages will be stored
```
`s/ -/:/` this line and the two following

jenswikl commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +450,125 @@ void optee_disable_shm_cache(struct optee *optee)
+}
+
+int optee_shm_register(struct tee_context *ctx, struct tee_shm *shm,
+		struct page **pages, size_t num_pages)
```
Alignment should match open parenthesis

jenswikl commented on drivers/tee/optee/rpc.c:
```diff
@@ -395,7 +447,8 @@ static void handle_rpc_func_cmd(struct tee_context *ctx, struct optee *optee,
  *
```
Description of `call_ctx` missing.

jenswikl commented on drivers/tee/optee/shm_pool.c:
```diff
@@ -0,0 +1,69 @@
+	.destroy_poolmgr = pool_op_destroy_poolmgr,
+};
+
+struct tee_shm_pool_mgr *optee_shm_pool_alloc_pages(void)
```
This function could have a description

jenswikl commented on drivers/tee/optee/shm_pool.c:
```diff
@@ -0,0 +1,69 @@
+}
+
+static void pool_op_free(struct tee_shm_pool_mgr *poolm,
+			     struct tee_shm *shm)
```
Alignment should match open parenthesis

jenswi-linaro reviewed (commented):
Some more comments after running kernel_patch_verify (https://github.com/nmenon/kernel_patch_verify.git)

"optee: enable dynamic SHM support"
Possible unwrapped commit description (prefer a maximum 75 chars per line)

"tee: use reference counting for tee_context"
WARNING: 'untill' may be misspelled - perhaps 'until'?

lorc reviewed (commented):
Thank you for review. I'm pushing review patches.
Comments for commit messages were fixed in place. I hope, this will not make GH crazy.

lorc committed:
[NEW SHM]tee: shm: inline tee_shm getter functions

Now, when struct tee_shm is defined in public header,
we can inline small getter functions.

Signed-off-by: Volodymyr Babchuk <vlad.babchuk@gmail.com>

lorc commented:
I also pushed new patch that inlines `tee_shm_is_registered()` and `tee_shm_get_id()`. 

I'm not sure if I have to inline another small functions like `tee_shm_put()` or `tee_shm_get_pa()`/`tee_shm_get_va()`. Please let me know, if I should include them in the patch.

jenswi-linaro commented:
Further inlining can wait until after this has been merged.

Looks good to me. If @etienne-lms is OK too, please squash and rebase to make it ready for merge.

Keep in mind that this is only the first review. As second review is expected to take place at the kernel mailing lists. During that review we'll keep stuff up to date on https://github.com/linaro-swg/linux/tree/optee

etienne-lms commented on include/linux/tee_drv.h:
```diff
@@ -211,6 +318,29 @@ void *tee_get_drvdata(struct tee_device *teedev);
+ */
+struct tee_shm *tee_shm_register(struct tee_context *ctx, unsigned long addr,
+				 size_t length, u32 flags);
+/**
```
minor: add empty line

etienne-lms reviewed (commented):
2 more minor comments.

lorc commented:
Hello. I have applied last comments by @etienne-lms, squashed, rebased and uploaded patch series.
Apart that I added static assert and `TODO: ` comment to the patch 
https://github.com/linaro-swg/linux/pull/29/commits/0aef7906585cf11daabdc6849646c60d6fa5b66d#diff-f4d6b53ce7cf409265b04a53ae7d121bR467 (tee: optee: add page list manipulation functions):
```
	/* TODO: add support for RichOS page sizes that != 4096 */
	BUILD_BUG_ON(PAGE_SIZE != OPTEE_MSG_NONCONTIG_PAGE_SIZE);
```

lorc commented:
Hello. So, what is next?

jockebech commented:
Hi @lorc , I think we should merge this on our linaro-swg/linux/optee branch, but then as Jens suggested, next thing is to send it to the kernel mailinglist, since that is where we like this to end up in the end. @etienne-lms @jenswi-linaro and other comments left?

lorc commented:
Hi @jbech-linaro, sure, I'll submit this patch series to LKML. I just want to be sure that you guys are okay with the patches.

jenswi-linaro commented:
1. Merge https://github.com/OP-TEE/optee_os/pull/1631
2. Merge https://github.com/linaro-swg/linux/pull/29 (with a "[NEW SHM]" prefix in the title on all commits to make it easy to identify the commits which are part of this and should be upstreamed)
3. Merge https://github.com/OP-TEE/optee_client/pull/72

Once that's done it's time for @lorc to send the linux patches to LKML and take another round of review there. During the review process we'll keep the testable setup in our development branch (`optee`) which we'll update with delta patches as needed.

jforissier commented:
+1 with what @jenswi-linaro said. Let me add action items:

> 1. Merge OP-TEE/optee_os#1631

**Action @lorc** : rebase onto current master, squash some commits together, apply R-b tags. See also my comment about rephrasing some commit subjects (but feel free to ignore them if it turns out I did not understand correctly!).
I'll merge the PR as soon as this is done.

> 2. Merge #29 (with a "[NEW SHM]" prefix in the title on all commits to make it easy to identify the commits which are part of this and should be upstreamed)

**Action @lorc** : add the "[NEW SHM]" prefix.
@jenswi-linaro : should A-b or R-b be added at this point?

> 3. Merge OP-TEE/optee_client#72

**Action @lorc** : address pending comments, rebase onto master (resolve conflicts).
Then we will gather the A-b/R-b and merge.

This is a cool feature and I'd really like to see it merged soon! So let's finalize this stuff :)
Thanks.

etienne-lms commented on include/linux/tee_drv.h:
```diff
@@ -394,6 +394,26 @@ void *tee_shm_get_va(struct tee_shm *shm, size_t offs);
+ */
+static inline ssize_t tee_shm_get_size(struct tee_shm *shm)
+{
+	return shm->size;
```
`sh->size` is a `size_t,` this returns a `ssize_t`. Likely to not be an issue, yet `tee_shm_get_size()` could simply return a `size_t`.

etienne-lms commented on include/linux/tee_drv.h:
```diff
@@ -404,6 +404,20 @@ static inline ssize_t tee_shm_get_size(struct tee_shm *shm)
 }
 
 /**
+ * tee_shm_get_pages() - Get list of pages that hold shared buffer
```
minor: s/hold/holds/

jenswi-linaro commented:
> Action @lorc : add the "[NEW SHM]" prefix.
@jenswi-linaro : should A-b or R-b be added at this point?

No, I'd rather take that on LKML. In the end I expect that I'll send the outcome of the review as a separate pull request to arm-soc with my S-o-B added to all commits.

etienne-lms commented on drivers/tee/optee/optee_msg.h:
```diff
@@ -67,11 +67,32 @@
+ * Every entry in buffer should point to a 4k page beginning (12 least
+ * significant bits must be equal to zero).
+ *
+ * 12 least significant of optee_msg_param.u.tmem.buf_ptr should hold page
```
minor: " * 12 least significant **bits** of optee_msg_param.u.tmem.buf_ptr should hold page
offset of **the** user buffer."

etienne-lms commented on drivers/tee/optee/call.c:
```diff
@@ -447,3 +448,50 @@ void optee_disable_shm_cache(struct optee *optee)
+
+static size_t get_pages_array_size(size_t num_entries)
+{
+	/* Number of user pages + number of pages to hold list of user pages  */
```
Minor: dual space at comment end

etienne-lms commented on drivers/tee/optee/call.c:
```diff
@@ -495,3 +495,67 @@ void optee_free_pages_array(void *array, size_t num_entries)
+int optee_shm_register(struct tee_context *ctx, struct tee_shm *shm,
+		       struct page **pages, size_t num_pages)
+{
+	struct tee_shm *shm_arg = NULL;
```
minor: useless init (although this change is overridden by a later change)

etienne-lms commented on drivers/tee/optee/core.c:
```diff
@@ -98,13 +98,79 @@ int optee_from_msg_param(struct tee_param *params, size_t num_params,
+			if (p->u.memref.size) {
+				size_t o = p->u.memref.shm_offs +
+					   p->u.memref.size;
+				if (o > tee_shm_get_size(shm))
```
minor: add empty line above

etienne-lms commented on drivers/tee/optee/core.c:
```diff
@@ -446,22 +452,49 @@ optee_config_shm_memremap(optee_invoke_fn *invoke_fn, void **memremaped_shm)
+			goto err_memunmap;
+		priv_mgr = rc;
+	} else {
+		size_t sz = OPTEE_SHM_NUM_PRIV_PAGES * PAGE_SIZE;
```
`const`

etienne-lms commented on include/linux/tee_drv.h:
```diff
@@ -42,11 +43,17 @@ struct tee_shm_pool;
  * @teedev:	pointer to this drivers struct tee_device
  * @list_shm:	List of shared memory object owned by this context
  * @data:	driver specific context data, managed by the driver
+ * @registered: reference counter for this structure
```
s/registered/refcount/

etienne-lms reviewed (commented):
I had a new review round to the patches. Please find the minor stylish findings...

Comment of commit "tee: shm: add accessors for buffer size and page offset"
**~~This~~These** two function**s**

Comment of commit "tee: shm: add page accessor functions "
... function to return**~~s~~** list ...

Comment of commit  "tee: optee: add page list manipulation functions"
 **~~This~~These** functions ...

Comment of commit: "tee: optee: add shared buffer registration functions"
* ... can use **~~this~~these** functions to **(un)**register ...
* Please note **~~,~~** that while _(remove coma)_ those functions **~~was~~were** added to optee code ...

Comment of commit "tee: optee: add optee-specific shared pool implementation"
* This is **a** simple pool that uses kernel page allocator. This pool can be used in case **~~if~~** OP-TEE supports dynamic shared memory.

lorc commented:
Thank you for review, Etienne. I addressed your comments right in appropriate patches. 
Also I added `[NEW SHM]` prefix to whole series.

lorc commented:
If there are no other comments on this PR, I'm going to post this patches to LKML later today.

lorc commented:
Ah, this is not so easy, as I expected :) It will take some time to resolve conflicts with master branch.
@jenswi-linaro, there are your `S-o-b` tag on your patches, that I'm going to submit as a part of the series.
But `tee: add register user memory` was reworked by me. Do you wish me to preserve your `S-o-b` tag?

(Honestly, it is the first time, I'm submitting сo-authored patches, I don't quite what is the correct procedure).

jenswi-linaro commented:
@lorc, I'm OK either way with `tee: add register user memory`, I guess I would have kept all `S-o-b`s.
